index.js

```diff
@@ -11,30 +10,54 @@ module.exports = {
   init (ssb, config) {
     const allowedTypes = getAllowedTypes(ssb, config)
 
-    ssb.publish.hook((publish, args) => {
+    const publishHook = ({ db2 }) => (publish, args) => {
       const [input, cb] = args
 
-      if (
-        isEncrypted(input) ||
-        hasRecps(input) ||
-        allowedTypes.has(input.type)
-      ) return publish(input, cb)
-
-      if (isAllowPublic(input)) {
+      if (db2 ? input.allowPublic === true : get(input, ['options', 'allowPublic']) === true) {
+        // allowPublic and has recps, disallowed
         if (hasRecps(input.content)) {
           return cb(new Error('recps-guard: should not have recps && allowPublic, check your code'))
         }
-        return publish(input.content, cb)
+
+        // allowPublic and no recps, allowed
+        return publish(db2 ? input : input.content, cb)
+      } else {
+        // without allowPublic, content isn't nested with db1 publish
+        const content = db2 ? input.content : input
+
+        // no allowPublic and has recps/can publish without recps, allowed
+        if (
+          (db2 ? (input.encryptionFormat !== undefined) : false) ||
+          isString(content) ||
+          hasRecps(content) ||
+          allowedTypes.has(content.type)
+        ) return publish(db2 ? input : content, cb)
+
+        // no allowPublic and no recps, disallowed
+        return cb(new Error(`recps-guard: public messages of type "${content.type}" not allowed`))
+      }
+    }
+
+    if (ssb.publish) {
```

Comment:
I think this should be 
```suggestion
    const isDB2 = Boolean(ssb.db)
    if (!isDB2) {
```

because `ssb.publish` could be a compat function which just ends up at `ssb.db.create` anyway

Reply:
yeah but we wouldn't be passing along the allowPublic option right? my understanding is the recps-guard hook is supposed to run first so we can turn the args normal

Reply:
like, a large part of the problem here i think stems from ssb.publish having one format for its args normally, but another one *if* recps-guard is installed and *if* the args haven't been processed by the recps-guard hook already. here we process the args, but if we then pass stuff along, we don't know if we're passing it along to normal publish or to compat db2 publish.